[FIX]: sort by spectral ratio part of fit + returned_filter not used

vpeterson opened the pull request:
This PR adds some more post-merged fixes to #8527 and #7070 [cc @dengemann]

Main issue: the option sort_by_spectral ratio was called within the` ssd.transform ` function when the sorting index should be learned when fitting the model.

Minors: 

1. The returned_filtered input was initialized but never used within the API. For using ssd in a decoding pipeline it would be useful to returned the projected data already filtered in the freq. band on interest.
2.  In `ssd.apply` components pick should be done after components sorting.

dengemann commented:
@vpeterson something isn't right here. I see many old commits. I guess you need to start from a clean new branch on top of the latest master.

vpeterson commented:
mhm, I've realized that @dengemann ! But naively thought that could be for calling the issue by #. it is weird because I created the branch on a supposed lastest master version.. should I rebase then? close the PR and start over?

dengemann commented:
yes, try to get a clean diff. Look at the preview before submitting the PR.

vpeterson commented:
Ok now seems to have done its job the rebase! @dengemann
Still some checking issues to solve, but I want also to show here some images showing the impact of having the sorting function called in the `ssd.transform` in instead of in the `ssd.fit` (decoding sklearn pipeline ssd+csp+lda, binary MI classification)

**Now:**
![scenario3_dataset2_sort_true_solved](https://user-images.githubusercontent.com/26253001/101956029-a2331500-3bdd-11eb-81dc-7adf0897f1f0.png)

**Before:**
![scenario3_dataset2_sort_true](https://user-images.githubusercontent.com/26253001/101956033-a4956f00-3bdd-11eb-8cc9-3537a53a99a5.png)

dengemann commented:
Cool! I still see some old commits in the history. Can you find a way to remove them and present the proposed change in one fresh commit?

vpeterson commented:
> Cool! I still see some old commits in the history. Can you find a way to remove them and present the proposed change in one fresh commit?

I did it now, but I guess I took one commit more than what I should. squash learning...

dengemann commented:
@vpeterson the history is still not clean it seems. You can see on Github that some commits include other authors, which as far as I understand is not intended. Can you see once more if your rebasing and sqashing worked?

agramfort commented:
if the diff is clean it's ok. We merge by squashing now

>

vpeterson commented:
still not able to solve it, but I will continue working tomorrow on this. The problem was that I accidentally squashed one commit more than what I should have done. Anyhow, @agramfort the tree is clean.

dengemann commented:
Ok I think we can move to reviewing. I saw many lines flagged by coverage CIs. We can probably add more tests.

vpeterson commented:
> Ok I think we can move to reviewing. I saw many lines flagged by coverage CIs. We can probably add more tests.

OK. I apparently clean the commit history, and added some more tests. But what is the py3 error about?

vpeterson commented:
Is there something I can do for solving the not successfully checks? @dengemann @agramfort ? By looking at the info, it seems to be a deprecation error in the pytest. I ask because I will not have access to my computer tomorrow, and if there is something that might need to be done now, I can devote time to that. 

FYI, I have cleaned the history (rebase and squash) and added some more tests to `test_ssd.py`. The pytests are successful when run on my computer.

dengemann commented on mne/decoding/ssd.py:
```diff
@@ -278,10 +282,6 @@ def apply(self, X):
-            _, sorter_spec = self.get_spectral_ratio(ssd_sources=X_ssd)
-
-        pick_patterns = self.patterns_[sorter_spec, :self.n_components].T
+        pick_patterns = self.patterns_[self.sorter_spec][:self.n_components].T
```
why change the square brackets? Wouldn't the old brackets have worked after changing for self.sorter_spec ?

vpeterson replied:
the problem here was that since the first dimension is where the components are, we need first to sort them and then pick the components. I noticed this when testing `ssd.apply` with `n_components ! = n_channels`. In addition, it goes along how the `ssd.transform` is implemented (line 215/217)

dengemann commented on mne/decoding/tests/test_ssd.py:
```diff
@@ -135,13 +135,28 @@ def test_ssd():
+              n_components=None, sort_by_spectral_ratio=True)
+    ssd.fit(X)
+    sorter_spec2 = ssd.sorter_spec
+    assert all(sorter_spec == sorter_spec2)
```
Is this the test we need?
Does it test that sorting from fit is applied during transform [open question]?
Perhaps we could test it more directly like running everything without sorting and then taking the sorter from a second fit and setting it to the first before transforming, then running a third fit with sorting activated and then comparing the results.
Any thoughts?

vpeterson replied:
mhmm that could be one option. But I also think that we can test actually the bug itself. Like splitting data in training and testing, evaluating some performance, and check there is no drop off in such performance.

vpeterson replied:
Hi @dengemann  ! check how I have made it now! it is implemented in `test_ssd:test_sorting`. Let me know what do you think now.

dengemann commented:
Hello @vpeterson this looks clean. Just left a few comments/questions. Also see that the coverage is complaining (see highlighted lines in diff). There are also a bunch of failing CIs. If we fix those and address we can probably merge.

vpeterson commented:
Sorry for my silence here. I made some changes regarding the test coverage. The unsuccessful checks seem to come from importing dependencies problems. Any idea how should I solve that or what is going on? @dengemann

dengemann commented:
@vpeterson @larsoner @agramfort it seems the failing tests are random. The changes look good to me, implementing the right fix. A second review would be good to have.

dengemann commented:
Example + tests run on my machine.

vpeterson commented:
> Example + tests run on my machine.

Hi @dengemann ! geat you tried the examples. They are also working on my computer, as well as the tests. 
Suggestions on how to proceed? should I commit a little change and see if the random errors are magically away?

drammock commented:
@vpeterson could you try closing and then re-opening this PR?  Usually that will re-start the CI servers...  some of them are having weird failures and I'm having trouble manually re-starting them.

drammock commented:
(you don't need to start a new PR, just click the "close" button and then the "re-open" button)

vpeterson commented:
Ok, I did the "trick" of closing and opening, but the errors are still there. I realized there are some warnings (deprecation warnings) that I am not 100% sure if I should fix them or they are more related to the main branch. 

These are the warnings I got after calling with pytest

`============================== warnings summary ===============================
..\..\..\anaconda3\envs\mnedev\lib\site-packages\pyreadline\py3k_compat.py:8
  c:\users\vpeterson\anaconda3\envs\mnedev\lib\site-packages\pyreadline\py3k_compat.py:8: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working
    return isinstance(x, collections.Callable)

..\..\..\anaconda3\envs\mnedev\lib\site-packages\numba\core\types\__init__.py:108
  c:\users\vpeterson\anaconda3\envs\mnedev\lib\site-packages\numba\core\types\__init__.py:108: DeprecationWarning: `np.long` is a deprecated alias for `np.compat.long`. To silence this warning, use `np.compat.long` by itself. In the likely event your code does not need to work on Python 2 you can use the builtin `int` for which `np.compat.long` is itself an alias. Doing this will not modify any behaviour and is safe. When replacing `np.long`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    long_ = _make_signed(np.long)

..\..\..\anaconda3\envs\mnedev\lib\site-packages\numba\core\types\__init__.py:109
  c:\users\vpeterson\anaconda3\envs\mnedev\lib\site-packages\numba\core\types\__init__.py:109: DeprecationWarning: `np.long` is a deprecated alias for `np.compat.long`. To silence this warning, use `np.compat.long` by itself. In the likely event your code does not need to work on Python 2 you can use the builtin `int` for which `np.compat.long` is itself an alias. Doing this will not modify any behaviour and is safe. When replacing `np.long`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    ulong = _make_unsigned(np.long)

mne\conftest.py:276
  C:\Users\vpeterson\Documents\GitHub\mne-python\mne\conftest.py:276: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    def renderer(backend_name, garbage_collect):

mne\conftest.py:287
  C:\Users\vpeterson\Documents\GitHub\mne-python\mne\conftest.py:287: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    def garbage_collect():

mne\conftest.py:303
  C:\Users\vpeterson\Documents\GitHub\mne-python\mne\conftest.py:303: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    def renderer_interactive(backend_name_interactive):

-- Docs: https://docs.pytest.org/en/stable/warnings.html
- generated xml file: C:\Users\vpeterson\Documents\GitHub\mne-python\junit-results.xml -
`

larsoner commented:
This probably needs a rebase or merge with `main`. Let me know if you want me to push a merge to help it build

vpeterson commented:
> This probably needs a rebase or merge with `main`. Let me know if you want me to push a merge to help it build

aaah I see! I can do the rebase and update of the PR. I will try it now, and see what happens then

vpeterson commented:
> > This probably needs a rebase or merge with `main`. Let me know if you want me to push a merge to help it build
> 
> aaah I see! I can do the rebase and update of the PR. I will try it now, and see what happens then

@larsoner, could you please push the merge? I have some issues with git at the moment (new computer, not everything working as it should be yet)

vpeterson commented:
Ok, there still are some errors that have nothing to do with branch/PR. Any idea how to proceed?

larsoner commented:
Pushed a little commit to run all CIs instead of just some. With this only the two `pip pre` jobs, one on Linux and one on Windows, should fail (this is our current status on `main`).

vpeterson commented:
Alright, we are getting closer. But besides I have added more tests, the codevov cannot be reached (and it is even worst now). By looking at the changed files it says that the returned_filtered is not checked, but I have made a whole function regarding this. What am I missing?

vpeterson commented:
All green @dengemann @larsoner !!!!!!

larsoner commented:
Thanks @vpeterson !
